implement cluster add and list

noam-codefresh opened the pull request:
still requires the current kube context to be set correctly before executing the command
might end up adding clusters in a different way completely

roi-codefresh commented on build/add-cluster-script.yaml:
```diff
@@ -0,0 +1,75 @@
+    environment:
+      - CHART_VERSION=${{CHART_VERSION}}
+    commands:
+      - echo "inside publish helm chart ${CHART_VERSION}"
```
so nothing here yet?

roi-codefresh commented on cmd/commands/cluster.go:
```diff
@@ -0,0 +1,258 @@
+	)
+
+	cmd := &cobra.Command{
+		Use:     "add RUNTIME_NAME --context CONTEXT_NAME",
```
should not have flags in the `USE` string, only positional params.
This should be an example.

roi-codefresh commented on cmd/commands/cluster.go:
```diff
@@ -0,0 +1,258 @@
+	cmd := &cobra.Command{
+		Use:     "add RUNTIME_NAME --context CONTEXT_NAME",
+		Short:   "Add a cluster to a given runtime",
+		Args:    cobra.MaximumNArgs(2),
```
Why 2 args? I see only one positional arg.

roi-codefresh commented on cmd/commands/cluster.go:
```diff
@@ -0,0 +1,258 @@
+	}
+
+	if runtime.RuntimeVersion == nil {
+		return fmt.Errorf("runtme \"%s\" has no version", opts.runtimeName)
```
```suggestion
		return fmt.Errorf("runtime \"%s\" has no version", opts.runtimeName)
```

roi-codefresh commented on cmd/commands/cluster.go:
```diff
@@ -0,0 +1,258 @@
+	}
+)
+
+var minimalSupportedVersion = semver.MustParse("99.99.99") // to be updated after app-proxy is ready
```
this is package scope, so not the best name

roi-codefresh commented on cmd/commands/cluster.go:
```diff
@@ -0,0 +1,258 @@
+	return opts.kubeFactory.Apply(ctx, "", manifests)
+}
+
+func createAddClusterKustomization(ingressUrl, contextName, server, csdpToken string, version *string) *kusttypes.Kustomization {
```
```suggestion
func createAddClusterKustomization(ingressUrl, contextName, server, csdpToken string, version string) *kusttypes.Kustomization {
```

Think it's better to dereference in the calling function

roi-codefresh commented on cmd/commands/cluster.go:
```diff
@@ -0,0 +1,258 @@
+func createAddClusterKustomization(ingressUrl, contextName, server, csdpToken string, version *string) *kusttypes.Kustomization {
+	resourceUrl := store.AddClusterDefURL
+	if strings.HasPrefix(resourceUrl, "http") && version != nil {
+		resourceUrl = fmt.Sprintf("%s?def=%s", resourceUrl, *version)
```
```suggestion
		resourceUrl = fmt.Sprintf("%s?ref=%s", resourceUrl, *version)
```

Is it supposed to be `ref` here?

roi-codefresh commented on hack/release.sh:
```diff
@@ -17,24 +17,18 @@ fi
+
 echo "uploading files:"
-ls -1a ./dist/*.tar.gz ./dist/*.sha256
+ls -1a ./dist/*.tar.gz ./dist/*.sha256 ./manifests/runtime.yaml
```
why do we suddenly need this?

roi-codefresh commented on manifests/add-cluster/add-cluster.sh:
```diff
@@ -0,0 +1,48 @@
+
+
+RESPONSE=$(curl -X POST ${INGRESS_URL}/app-proxy/api/clusters \
+  -H 'Content-Type: application/json' \
```
maybe you should configure here some sort of retry. pretty sure you can do that with a flag or two.

roi-codefresh commented on manifests/add-cluster/helm/templates/clusterrole.yaml:
```diff
@@ -0,0 +1,17 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: {{ .Values.serviceAccount | default "argocd-manager" }}-role
```
Need to concat chart unique value so those won't collide with other cluster scope resources.

roi-codefresh replied:
It's the `{{ include "csdp-add-cluster.fullname" . }}` stuff you have on a number of resources already.

roi-codefresh commented on manifests/add-cluster/helm/templates/clusterrolebinding.yaml:
```diff
@@ -0,0 +1,14 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: {{ .Values.serviceAccount | default "argocd-manager" }}-role-binding
```
here too

roi-codefresh commented on manifests/add-cluster/helm/templates/job.yaml:
```diff
@@ -0,0 +1,55 @@
+  labels:
+    {{- include "csdp-add-cluster.labels" . | nindent 4 }}
+spec:
+  ttlSecondsAfterFinished: 600 # stick around for 10m
```
where is a retry policy?

roi-codefresh commented on manifests/add-cluster/helm/values.yaml:
```diff
@@ -0,0 +1,18 @@
+serviceAccountName: argocd-manager
+
+# ingressUrl is the url used to access the CSDP runtime
+ingressUrl: <ingressUrl>
```
Is this the base url? or need the whole path? Maybe an example?

roi-codefresh commented on manifests/add-cluster/helm/values.yaml:
```diff
@@ -0,0 +1,18 @@
+# ingressUrl is the url used to access the CSDP runtime
+ingressUrl: <ingressUrl>
+
+# csdoToken is used to authenticate against the app-proxy
```
```suggestion
# csdpToken is used to authenticate against the app-proxy
```

roi-codefresh replied:
I don't think it's a normal thing a user would like a cli thing to do, to show his secrets. But we should have a command to generate a populated helm values file with all of the values.

roi-codefresh commented on manifests/add-cluster/kustomize/configmap.yaml:
```diff
@@ -0,0 +1,9 @@
+data:
+  ingressUrl: <ingressUrl>
+  contextName: <contextName>
+  server: <server
```
```suggestion
  server: <server>
```

roi-codefresh commented on manifests/add-cluster/kustomize/job.yaml:
```diff
@@ -0,0 +1,52 @@
+      restartPolicy: Never
+      containers:
+        - name: main
+          image: quay.io/codefresh/csdp-add-cluster:0.1.0
```
put latest here and replace in the `kustomization.yaml` file

roi-codefresh commented on manifests/add-cluster/kustomize/secret.yaml:
```diff
@@ -0,0 +1,7 @@
+  name: csdp-add-cluster-secret
+type: Opaque
+data:
+  csdpToken: <csdpToken>
```
Maybe we need to think about some way to clean those after the job is done. But not crucial for now.

roi-codefresh replied:
gr8

roi-codefresh commented on pkg/util/cd/util.go:
```diff
@@ -39,44 +41,55 @@ type (
+}
+
+func GetClusterList(ctx context.Context, kubeContext, namespace string, insecure bool) (*v1alpha1.ClusterList, error) {
+	argoClient, err := createArgoClient(kubeContext, namespace, insecure)
```
Do we plan to move this logic and use the platform client to list the clusters? Because I don't think it's a good idea to read those from argocd directly. It's different from all the rest of the entities.

ATGardner committed:
implement cluster add and list

added kustomization, fixed helm

added kustimized add-cluster to release

generate kustomization for add-cluster

create and build kustomization

added dry-run

added helm labels

renamed add-cluster to csdp-add-clsuter

added add-cluster-script pipeline

added runtime version check

fixed secret creation, delete secret on success

keep the job for 10m after completion

add helm release.name to resources

added argocd vars to app-proxy deployment

renamed pipeline script file

added the default argocd username into the configmapGenerator

updated version to 0.0.283

deploy app-proxy 1.923.0
